### PR TITLE
feat: introduce an intermediate version that is compatible with both …

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

#16049 can be rolled back to this PR, and this PR can be rolled back to older versions. #16049 cannot be directly rolled back to versions before this PR.

`v1_main`: 1.2.630-nightly-fb9fc9a680 (Latest version of the main branch at the time of testing)

`v2_this_pr`: This pull request

`v3_vacuum2_pr`: Pull request #16049 (Implementing vacuum2 functionality)
### Test snapshot_location_from_uuid()
Execute the following statements in `v3_vacuum2_pr`:
```
create or replace table t1(a int);
insert into t1 values (1);
insert into t1 values (2);
insert into t1 values (3);
```

Change databend query to `v1_main`, and execute:
```
root@localhost:8000/default> create or replace stream s1 on table t1 at (snapshot => '0191c068a10b7ba8954393ccd7f819ce');
error: APIError: ResponseError with 2013: No historical data found at given point
```
Since `v1_main` is not compatible with `v3_vacuum2_pr`, the error is expected, and this indicates that downgrade from `v3_vacuum2_pr` to `v1_main` is not applicable, if new data has been written by using  `v3_vacuum2_pr`.

Change databend query to `v2_this_pr`, and execute:
```
root@localhost:8000/default> create or replace stream s1 on table t1 at (snapshot => '0191c068a10b7ba8954393ccd7f819ce');

CREATE
OR REPLACE stream s1 ON TABLE t1 at (snapshot = > '0191c068a10b7ba8954393ccd7f819ce')

0 row written in 0.167 sec. Processed 0 row, 0B (0 row/s, 0B/s)
```
This shows that this PR `v2_this_pr` is compatible with `v3_vacuum2_pr`: After upgraded to `v3_vacuum2_pr`, and wrote down some new data, it still possible to downgrade to `v2_this_pr`.

### Test block_id_from_location()
Execute the following statements in `v3_vacuum2_pr`:
```
create or replace table t1(a int, b int);
create or replace table t2(a int, b int);
create or replace stream s2 on table t1 append_only = true;
insert into t2 values(2,2),(3,3),(4,4);
insert into t1 values(1,1),(2,3),(3,3);

```
Change databend query to `v1_main`, and execute:
```
root@localhost:8000/default> merge into t1 using t2 on t1.a=t2.a when matched then update set t1.b=t2.b when not matched then insert *;
error: APIError: ResponseError with 1001: invalid character: expected an optional prefix of `urn:uuid:` followed by [0-9a-fA-F-], found `g` at 1
```

Change databend query to `v2_this_pr`, and execute:
```
root@localhost:8000/default> merge into t1 using t2 on t1.a=t2.a when matched then update set t1.b=t2.b when not matched then insert *;

MERGE INTO t1 USING t2 ON t1.a = t2.a
WHEN matched THEN
UPDATE
SET
  t1.b = t2.b
  WHEN NOT matched THEN
INSERT
  *

-[ RECORD 1 ]-----------------------------------
number of rows inserted: 1
 number of rows updated: 2

1 row read in 1.320 sec. Processed 6 row, 47B (4.55 rows/s, 35B/s)
```

## Tests

- [ ] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [x] Compatible test

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16354)

### CODE REVIEW

File: src/query/storages/fuse/src/io/locations.rs [hunk 1]
```diff
@@ -170,7 +172,7 @@ impl TableMetaLocationGenerator {
         let splits = loc.split('/').collect::<Vec<_>>();
         let len = splits.len();
         let prefix = splits[..len - 2].join("/");
-        let block_name = splits[len - 1];
+        let block_name = trim_vacuum2_object_prefix(splits[len - 1]);
```
Comment (dantengsky) [inline src/query/storages/fuse/src/io/locations.rs:175]: do we need to trim the 'g' in this case?

File: src/query/storages/fuse/src/io/locations.rs [hunk 2]
```diff
@@ -182,7 +184,7 @@ impl TableMetaLocationGenerator {
...
```
Comment (dantengsky) [inline src/query/storages/fuse/src/io/locations.rs:187]: do we need to trim the 'g' in this case?

File: src/query/storages/fuse/src/io/locations.rs (hunk 1 above)
Comment (SkyFan2002) [reply src/query/storages/fuse/src/io/locations.rs:175]: not need

File: src/query/storages/common/table_meta/src/meta/utils.rs [hunk 3]
```diff
@@ -78,3 +80,30 @@ pub fn parse_storage_prefix(options: &BTreeMap<String, String>, table_id: u64) -
     }
     Ok(prefix)
 }
+
+#[inline]
+pub fn trim_vacuum2_object_prefix(key: &str) -> &str {
```
Comment (BohuTANG) [inline src/query/storages/common/table_meta/src/meta/utils.rs:85]: Better to add a unit test for this func :)

**Review by dantengsky — APPROVED** [review]

**Comment by SkyFan2002** [comment]
https://github.com/datafuselabs/databend/pull/16397, @dantengsky

**Review by dantengsky — APPROVED** [review]

**Pull request merged**